util/settings/settings_test.go

```diff
@@ -420,3 +421,74 @@ func Test_validateExternalURL(t *testing.T) {
 		})
 	}
 }
+
+func TestSecretKeyRef(t *testing.T) {
+	data := map[string]string{
+		"dex.config": `
+      connectors:
+        # GitHub example
+        - type: github
+          id: github
+          name: GitHub
+          config:
+            clientID: aabbccddeeff00112233
+            clientSecret: $google:clientSecret
```

Comment:
Sorry, I'm having a hard time understanding the "why" of this PR :-) 

Are we switching from `$google.clientSecret` to `$google:clientSecret` ?

Reply:
Hello. The _why_ is https://github.com/argoproj/argo-cd/issues/4188 (as referred to the title of current PR).

To summarise, one would refer to secret that live _outside_ of  _argocd-secret_ K8S `Secret`'s data.

I'm not switching, rather adding another syntax.

Explanation by example (the one in `settings_test` is not obvious apparently), to clear the things out:
 - As defined (and limited by) in current documentation:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: argocd-secret
  namespace: argocd
  labels:
    app.kubernetes.io/name: argocd-secret
    app.kubernetes.io/part-of: argocd
type: Opaque
data:
  ...
  # Store client secret like below.
  # Ensure the secret is base64 encoded
  oidc.auth0.clientSecret: <client-secret-base64-encoded>
  ...
...
apiVersion: v1
kind: ConfigMap
metadata:
  name: argocd-cm
  namespace: argocd
  labels:
    app.kubernetes.io/name: argocd-cm
    app.kubernetes.io/part-of: argocd
data:
  ...
  oidc.config: |
    name: Auth0
    clientID: aabbccddeeff00112233
    # Reference key in argocd-secret
    clientSecret: $oidc.auth0.clientSecret
```
so in `clientSecret: $oidc.auth0.clientSecret`, `oidc...` **should** be in `argocd-secret`'s `data`, right ?

 - Now (after this PR, I mean), we can **also** have
 ```yaml
apiVersion: v1
kind: Secret
metadata:
  name: another-secret # that is not _handled_ by Argocd, and eventually synced by tools such as External-secret
data:
  ...
  # Store client secret like below.
  # Ensure the secret is base64 encoded
  oidc.auth0.clientSecret: <client-secret-base64-encoded>
...
apiVersion: v1
kind: ConfigMap
metadata:
  name: argocd-cm
data:
  ...
  oidc.config: |
    name: Auth0
    clientID: aabbccddeeff00112233
    # Reference key in argocd-secret
    clientSecret: $another-secret:oidc.auth0.clientSecret
```
where `oidc.auth0.clientSecret` is in `data` field of another K8S secret.

I was personally motivated to answer this issue because we have tool that sync secrets with external secret management tool (such as Vault or AWS secret manager) and periodically updates the whole K8S `Secret` which have the effect to be override the content of `argocd-secret` that `ArgoCD` updates as well (with its admin password, its certificates...)  
So, yes, it indeed lakes of some documentation; that I'm gonna add.

Reply:
This makes sense, thank you very much!